src/Model/Workspace.re

```diff
@@ -8,11 +8,12 @@
  * - Per-workspace configuration
  */
 
-type workspace = {
+type t = {
   workingDirectory: string,
   rootName: string,
 };
 
-type t = option(workspace);
-
-let initial: t = None;
+let initial = workingDirectory => {
+  workingDirectory,
+  rootName: Filename.basename(workingDirectory),
+};
```

Comment:
Workspace is no longer an `option`. I'm not sure why it ever was, as it's not necessary and was a pain in the neck to deal with everywhere. It was actually just `None` until the `FileExplorer` set it to the current working directory, which was a really odd place to make that decision and too late to properly deal with persistence, so now it's set when the state is created initially instead.

Reply:
Unfortunately, there are cases where we can't necessarily make a good decision on a 'default' workspace directory - for example, this bug on OSX: https://github.com/onivim/oni2/issues/619

By default, when you launch Onivim 2 on OSX, it uses `~` as the working directory... which has some nasty side effects as documented in https://github.com/onivim/oni2/issues/619. It's a pretty horrible first run experience on Mac, at the moment.

In that case - we're better off not setting a workspace, and giving the user an option to choose one (similar to how VSCode behaves). So we're going to have to represent the no-active-workspace case in some way (and in that case - we shouldn't pick up any persistence).

Reply:
> By default, when you launch Onivim 2 on OSX, it uses ~ as the working directory.

Is that because it is explicitly passed in as a CLI argument, or just because that's our fallback? If the latter, why not just use a different fallback? It would make our life a lot easier.

Either way, it was de facto always set before though, because of the file explorer.So to revert it we'll have to decide what to do with that when there's no workspace.

Reply:
> Is that because it is explicitly passed in as a CLI argument, or just because that's our fallback?

It's not expliclity passed, but I guess it's the default working directory OSX uses when opened via Finder.

> If the latter, why not just use a different fallback? It would make our life a lot easier.

We've tried to find a good fallback but have been unsuccessful - I think the best thing for us to do is to create an experience around 'no-workspace-yet' (ie, an 'empty' experience for the file explorer that prompts the user to open a workspace).

As you mentioned, we already had a 'default' workspace set anyway, so this isn't a step backward - but it is something we'll have to figure out a solution in order to improve the first-run experience on OSX and fix https://github.com/onivim/oni2/issues/619

Reply:
> We've tried to find a good fallback but have been unsuccessful 

`Documents` doesn't work? Or a dedicated oni2 workspace directory? There's some more discussion about this in #906 btw.

> I think the best thing for us to do is to create an experience around 'no-workspace-yet' 

That means we have to deal with the possibility of there not being a workspace everywhere, and therefore make a ton of decisions about what to do instead. What should we do with search if there's no workspace? What should we do with `:cd`? What should we do with `:e`? With `!`? Is no-oping or special casing everything that doesn't work without a workspace a better first-run experience?

> As you mentioned, we already had a 'default' workspace set anyway

Not just in the state though.There's always a working directory, and plenty of functionality uses that instead of the workspace. Being able to use `:e` and `:cd` without the UI giving you a clue about where you are certainly doesn't sound like a great experience.

> so this isn't a step backward

Buit it is, because we now persist the workspace. So the logic for that needs to be convoluted even more to account for the possibility of there being no workspace. Not a major step backwards, but it does make this and everything else much more complicated.

I really can;t see how this would end up being a better experience for anyone.

Reply:
At the high level, all I want is a fix for https://github.com/onivim/oni2/issues/619 - whatever that entails. If it means that we have to handle a `None` case, then that's what we have to do - it's simply not an acceptable user experience as it stands today. 

On the other hand, if we can find a solution that solves https://github.com/onivim/oni2/issues/619 with a good default, I'm all for that and of course I'd prefer to keep things simple 😄 

> `Documents` doesn't work?

That's the best suggestion so far, IMO! We can give that a shot, I believe it should fix https://github.com/onivim/oni2/issues/619 , without adding the complexity of handling the no-workspace case.

Reply:
Cool! Should we try that just for macOS then, or for all platforms? it does seem more natural for Windows as well, but on Linux I think there's a possibility that there is no `Documents` folder so we'll have to fall back to `$HOME`. Which is arguably a more natural choice on Linux anyway.

Reply:
I like the idea of using the documents folder, and then falling back to home if its not available (when starting via Finder / UI - when starting from the terminal, we can use that working directory).

Might be a good use case for the reason-native `dir` library: https://github.com/facebookexperimental/reason-native/tree/master/src/dir#diruser - They have a `Dir.User.document` API, that, if it is `None`, we could fall-back to `Dir.User.home`.

Sounds like we have a good plan, and shouldn't block this PR 👍  Thanks for brainstorming this!

Reply:
> when starting via Finder / UI - when starting from the terminal, we can use that working directory

How do we distinguish between being started from Finder/UI vs the terminal?

Reply:
I think that's the big issue for us to figure out... 

On OSX, there is a weird `-psn` argument (that we filter out):
https://github.com/onivim/oni2/blob/b3be9160e3a617c32a574f53588e53af0bc578c3/src/bin_launcher/Oni2.re#L97
 I'm not sure if this only happens on first launch (from gatekeeper), or if it happens anytime launched through Finder, but if it is the latter, that would be a reliable way on OSX.

Another option might be to check if the `SHELL` env variable is set - wouldn't be the case for launching from finder, but would be when running from the terminal...

Reply:
Another alternative might be to do like `vscode`, require a `.` argument to open the working directory from the terminal.

Reply:
That sounds good!

Reply:
> Might be a good use case for the reason-native dir library: https://github.com/facebookexperimental/reason-native/tree/master/src/dir#diruser 

I'm not sure how to use this. It doesn't seem to be released. And since it uses FFI I can't easily just copy it into out project either, which might also have licensing issues.

Reply:
Oh there's an `opam` file in the root I can probably use.

Reply:
Ah, ya - looks like we could add a dependency like this with the json file, as well:
```
    "@reason-native/dir": "facebookexperimental/reason-native:dir.json#bb80cbc"
```

Reply:
Oh, that works as well? That doesn't seem to be documented. Worked fine with the opam file though, but I had to add `Fp` as well. Which we were going to do anyway so...

Reply:
Done in https://github.com/onivim/oni2/pull/1703/commits/0eaedb31569816addebbdb3963706347d0eb4328. I'll mark this as fixing #619 too then.